### PR TITLE
fix(hogql): dateAdd should be date_add

### PR DESCRIPTION
**Pull request by EDsCODE**
## Problem

- dateAdd doesn't work
- https://posthoghelp.zendesk.com/agent/tickets/14752 (moved to PostHog: https://us.posthog.com/project/2/support/tickets/17066)

## Changes

- date_add does

👉 _Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review._

### CODE REVIEW

**Comment by mariusandra** [comment]
Questions.
1. Why? The PR doesn't say.
2. What will break?

File: posthog/hogql/functions/mapping.py [hunk 1]
```diff
@@ -462,7 +462,7 @@ def compare_types(arg_types: list[ConstantType], sig_arg_types: tuple[ConstantTy
     "age": HogQLFunctionMeta("age", 3, 3),
     "dateDiff": HogQLFunctionMeta("dateDiff", 3, 3),
     "dateTrunc": HogQLFunctionMeta("dateTrunc", 2, 2),
-    "dateAdd": HogQLFunctionMeta("dateAdd", 3, 3),
```
Comment (Gilbert09) [inline posthog/hogql/functions/mapping.py:465]: `dateAdd` is an alias if `date_add` - so I don't see why this wouldn't work. [docs](https://clickhouse.com/docs/en/sql-reference/functions/date-time-functions#date_add). I'd be keen to not change this if we can help it, it'll become one of the only funcs in snake case, and may break existing queries

Comment (Gilbert09) [reply posthog/hogql/functions/mapping.py:465]: image

**Comment by EDsCODE** [comment]
> Questions.
> 
> 1. Why? The PR doesn't say.
> 2. What will break?

1. Hmm, so just sifted through clickhouse source and found more examples. When the bug was first reported, I basically tried all variations of `dateAdd` and nothing worked even though documentation suggests something like `SELECT dateAdd('YEAR', 1, toDateTime('2022-01-01'))` would work but this didn't work in hogQL or in clickhouse at all. But it looks like `SELECT dateAdd(YEAR, 1, toDateTime('2022-01-01'))` is the syntax that works on clickhouse but not on hogQL because 'YEAR' token without quotes isn't supported.

I changed it to date_add because `SELECT date_add(toDateTime('2022-01-01'), INTERVAL 1 YEAR)` which was the only formulation that seemed to work with hogQL. I didn't realize what I just found out this morning about `dateAdd`, it seemed like it was entirely broken in clickhouse.

So, I guess with that understanding, the better solution is to keep the existing signature but make sure dateAdd supports 3 arguments or 2 (as suggested in docs) and for 3 arguments, the first argument can be a quoteless time interval token. @mariusandra 

2. From the support ticket it seems like nothing that isn't already? They tried just about all the formulations for `dateAdd` and nothing works

**Comment by mariusandra** [comment]
We definitely certainly absolutely do not want to *remove* functions from HogQL. This PR removes `dateAdd`. Did you check how many people use it in their insights, API queries and notebooks before proposing this? How many of them will have something that'll now break and ends up on the Product Analytics support queue?

The new function also does not look like any of the other functions we have. If you really want to write `date_add` in ClickHouse, change what `dateAdd` prints to, but there's no reason to change the signature in HogQL.

Will swapping `INTERVAL 1 YEAR` --> `toIntervalYear(1)` make a difference?

**Comment by EDsCODE** [comment]
> Did you check how many people use it in their insights, API queries and notebooks before proposing this? How many of them will have something that'll now break and ends up on the Product Analytics support queue?

No, because the function entirely doesn't work so I was working with the assumption that it's 0 and/or no one reported it. Just checked and here's the data (please check my query logic in case I might be missing how hogql queries are stored)

[1 on US](https://metabase.prod-us.posthog.dev/question#eyJkYXRhc2V0X3F1ZXJ5Ijp7InR5cGUiOiJuYXRpdmUiLCJuYXRpdmUiOnsicXVlcnkiOiJTRUxFQ1QgbmFtZSwgdGVzdCwgaWQsIHRlYW1faWQgRlJPTSAoXG4gICAgU0VMRUNUIGNhc3QocXVlcnktPidzb3VyY2UnLT4ncXVlcnknIGFzIHZhcmNoYXIpIEFTIHRlc3QsIGlkLCB0ZWFtX2lkLCBuYW1lIEZST00gcG9zdGhvZ19kYXNoYm9hcmRpdGVtIFdIRVJFIHF1ZXJ5IElTIE5PVCBOVUxMIEFORCBxdWVyeS0-J3NvdXJjZSctPidxdWVyeScgSVMgTk9UIE5VTEwgXG4pIGFzIHF1ZXJpZXMgV0hFUkUgdGVzdCBMSUtFICclZGF0ZUFkZCUnOjp0ZXh0IiwidGVtcGxhdGUtdGFncyI6e319LCJkYXRhYmFzZSI6MzR9LCJkaXNwbGF5Ijoic2NhbGFyIiwicGFyYW1ldGVycyI6W10sInZpc3VhbGl6YXRpb25fc2V0dGluZ3MiOnt9fQ==) (which noted that dateAdd doesn't work)

[0 on EU](https://metabase.prod-eu.posthog.dev/question#eyJkYXRhc2V0X3F1ZXJ5Ijp7InR5cGUiOiJuYXRpdmUiLCJuYXRpdmUiOnsicXVlcnkiOiJTRUxFQ1QgbmFtZSwgdGVzdCwgaWQsIHRlYW1faWQgRlJPTSAoXG4gICAgU0VMRUNUIGNhc3QocXVlcnktPidzb3VyY2UnLT4ncXVlcnknIGFzIHZhcmNoYXIpIEFTIHRlc3QsIGlkLCB0ZWFtX2lkLCBuYW1lIEZST00gcG9zdGhvZ19kYXNoYm9hcmRpdGVtIFdIRVJFIHF1ZXJ5IElTIE5PVCBOVUxMIEFORCBxdWVyeS0-J3NvdXJjZSctPidxdWVyeScgSVMgTk9UIE5VTEwgXG4pIGFzIHF1ZXJpZXMgV0hFUkUgdGVzdCBMSUtFICclZGF0ZUFkZCUnOjp0ZXh0IiwidGVtcGxhdGUtdGFncyI6e319LCJkYXRhYmFzZSI6MzR9LCJkaXNwbGF5IjoidGFibGUiLCJwYXJhbWV0ZXJzIjpbXSwidmlzdWFsaXphdGlvbl9zZXR0aW5ncyI6e319)



> The new function also does not look like any of the other functions we have. If you really want to write date_add in ClickHouse, change what dateAdd prints to, but there's no reason to change the signature in HogQL.

> Will swapping INTERVAL 1 YEAR --> toIntervalYear(1) make a difference?

Agreed the snake case doesn't make sense. I had just changed it to match clickhouse more closely. 

We can keep dateAdd but can we change the function to accept only 2 args? That would work. Right now with 3, it expects an interval token as the first argument like `YEAR` and it doesn't work. `INTERVAL 1 YEAR` isn't the issue. That argument works but you can't pass that into the 3 argument version of dateAdd. 

If we want to keep the signature exactly the same, I think we would need to support `YEAR | MONTH | WEEK | DAY | MINUTE` as its own AST node and valid in specific functions.

**Review by mariusandra — APPROVED** [review]
I'm more than happy to change the wrong number of arguments to the correct one 😄

**Review by Gilbert09 — APPROVED** [review]

**Pull request merged**